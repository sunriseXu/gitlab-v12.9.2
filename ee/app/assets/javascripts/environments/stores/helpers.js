/**
 * Deploy boards are EE only.
 *
 * @param {Object} environment
 * @returns {Object}
 */
// eslint-disable-next-line import/prefer-default-export
export const setDeployBoard = (oldEnvironmentState, environment) => {
  let parsedEnvironment = environment;
  if (environment.size === 1 && environment.rollout_status) {
    parsedEnvironment = Object.assign({}, environment, {
      hasDeployBoard: true,
      isDeployBoardVisible:
        oldEnvironmentState.isDeployBoardVisible === false
          ? oldEnvironmentState.isDeployBoardVisible
          : true,
      deployBoardData:
        environment.rollout_status.status === 'found' ? environment.rollout_status : {},
      isLoadingDeployBoard: environment.rollout_status.status === 'loading',
      isEmptyDeployBoard: environment.rollout_status.status === 'not_found',
      hasLegacyAppLabel: environment.rollout_status.has_legacy_app_label,
    });
  }
  return parsedEnvironment;
};
