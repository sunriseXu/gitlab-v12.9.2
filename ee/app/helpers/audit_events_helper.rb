# frozen_string_literal: true

module AuditEventsHelper
  def human_text(details)
    return details[:custom_message] if details[:custom_message]

    details.map { |key, value| select_keys(key, value) }.join(" ").humanize
  end

  def select_keys(key, value)
    if key =~ /^(author|target)_.*/
      ""
    elsif key.to_s == 'ip_address' && value.blank?
      ""
    elsif key =~ /^expiry_(from|to)$/ && value.blank?
      "#{key} <strong>never expires</strong>"
    else
      "#{key} <strong>#{value}</strong>"
    end
  end
end
