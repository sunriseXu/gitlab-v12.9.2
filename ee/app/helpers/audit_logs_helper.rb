# frozen_string_literal: true

module AuditLogsHelper
  def audit_entity_type_options
    [
      { id: 'All', text: 'All Events' },
      { id: 'Group', text: 'Group Events' },
      { id: 'Project', text: 'Project Events' },
      { id: 'User', text: 'User Events' }
    ]
  end

  def audit_entity_type_label(selected)
    selected = 'All' unless selected.present?

    audit_entity_type_options.find { |type| type[:id] == selected }[:text]
  end
end
