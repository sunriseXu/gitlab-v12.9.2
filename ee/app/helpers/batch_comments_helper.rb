# frozen_string_literal: true

module BatchCommentsHelper
  def batch_comments_enabled?
    current_user.present? && @project.feature_available?(:batch_comments, current_user)
  end
end
