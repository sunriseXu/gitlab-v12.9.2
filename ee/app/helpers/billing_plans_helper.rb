# frozen_string_literal: true

module BillingPlansHelper
  def subscription_plan_info(plans_data, current_plan_code)
    plans_data.find { |plan| plan.code == current_plan_code }
  end

  def number_to_plan_currency(value)
    number_to_currency(value, unit: '$', strip_insignificant_zeros: true, format: "%u%n")
  end

  def current_plan?(plan)
    plan.purchase_link&.action == 'current_plan'
  end

  def plan_purchase_link(href, link_text)
    if href
      link_to link_text, href, class: 'btn btn-success'
    else
      button_tag link_text, class: 'btn disabled'
    end
  end

  def new_gitlab_com_trial_url
    "#{EE::SUBSCRIPTIONS_URL}/trials/new?gl_com=true"
  end

  def subscription_plan_data_attributes(group, plan)
    return {} unless group

    {
      namespace_id: group.id,
      namespace_name: group.name,
      plan_upgrade_href: plan_upgrade_url(group, plan),
      customer_portal_url: "#{EE::SUBSCRIPTIONS_URL}/subscriptions"
    }
  end

  def plan_upgrade_url(group, plan)
    return unless group && plan&.id

    "#{EE::SUBSCRIPTIONS_URL}/gitlab/namespaces/#{group.id}/upgrade/#{plan.id}"
  end

  def plan_purchase_url(group, plan)
    "#{plan.purchase_link.href}&gl_namespace_id=#{group.id}"
  end

  def plan_feature_short_list(plan)
    return [] unless plan.features

    plan.features.sort_by! { |feature| feature.highlight ? 0 : 1 }[0...4]
  end

  def plan_purchase_or_upgrade_url(group, plan, current_plan)
    if group.upgradable?
      plan_upgrade_url(group, current_plan)
    else
      plan_purchase_url(group, plan)
    end
  end

  def show_plans?(namespace)
    namespace.trial_active? || !namespace.gold_plan?
  end

  def show_trial_banner?(namespace)
    return false unless params[:trial]

    root = namespace.has_parent? ? namespace.root_ancestor : namespace
    root.trial_active?
  end

  def namespace_for_user?(namespace)
    namespace == current_user.namespace
  end
end
