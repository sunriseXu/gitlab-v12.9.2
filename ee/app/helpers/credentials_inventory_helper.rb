# frozen_string_literal: true

module CredentialsInventoryHelper
  VALID_FILTERS = %w(ssh_keys personal_access_tokens).freeze

  def show_personal_access_tokens?
    return true if params[:filter] == 'personal_access_tokens'

    VALID_FILTERS.exclude? params[:filter]
  end

  def show_ssh_keys?
    params[:filter] == 'ssh_keys'
  end

  def credentials_inventory_feature_available?
    License.feature_available?(:credentials_inventory)
  end

  def credentials_inventory_path(args)
    raise NotImplementedError, "#{self.class} does not implement #{__method__}"
  end

  def user_detail_path(user)
    raise NotImplementedError, "#{self.class} does not implement #{__method__}"
  end
end
