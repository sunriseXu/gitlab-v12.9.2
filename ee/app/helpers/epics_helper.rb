# frozen_string_literal: true

module EpicsHelper
  def epic_show_app_data(epic)
    EpicPresenter.new(epic, current_user: current_user).show_data(author_icon: avatar_icon_for_user(epic.author), base_data: issuable_initial_data(epic))
  end

  def epic_endpoint_query_params(opts)
    opts[:data] ||= {}
    opts[:data][:endpoint_query_params] = {
      only_group_labels: true,
      include_ancestor_groups: true,
      include_descendant_groups: true
    }.to_json

    opts
  end

  def epic_state_dropdown_link(state, selected_state)
    link_to epic_state_title(state), page_filter_path(state: state), class: state == selected_state ? 'is-active' : ''
  end

  def epic_state_title(state)
    titles = {
      "opened" => "Open"
    }

    _("%{state} epics") % { state: (titles[state.to_s] || state.to_s.humanize) }
  end

  def epic_timeframe(epic)
    short_format = '%b %d'
    long_format = '%b %d, %Y'

    if epic.start_date.present? && epic.end_date.present?
      start_date_format = epic.start_date.year == epic.end_date.year ? short_format : long_format

      "#{epic.start_date.strftime(start_date_format)} – #{epic.end_date.strftime(long_format)}"
    elsif epic.start_date.present?
      s_('GroupRoadmap|%{dateWord} – No end date') % { dateWord: epic.start_date.strftime(long_format) }
    elsif epic.end_date.present?
      s_("GroupRoadmap|No start date – %{dateWord}") % { dateWord: epic.end_date.strftime(long_format) }
    end
  end
end
