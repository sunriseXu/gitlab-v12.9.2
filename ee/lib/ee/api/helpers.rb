# frozen_string_literal: true

module EE
  module API
    module Helpers
      extend ::Gitlab::Utils::Override
      include ::Gitlab::Utils::StrongMemoize

      def require_node_to_be_enabled!
        forbidden! 'Geo node is disabled.' unless ::Gitlab::Geo.current_node&.enabled?
      end

      def gitlab_geo_node_token?
        headers['Authorization']&.start_with?(::Gitlab::Geo::BaseRequest::GITLAB_GEO_AUTH_TOKEN_TYPE)
      end

      def authenticate_by_gitlab_geo_node_token!
        unauthorized! unless authorization_header_valid?
      rescue ::Gitlab::Geo::InvalidDecryptionKeyError, ::Gitlab::Geo::InvalidSignatureTimeError => e
        render_api_error!(e.to_s, 401)
      end

      def check_gitlab_geo_request_ip!
        unauthorized! unless ::Gitlab::Geo.allowed_ip?(request.ip)
      end

      override :current_user
      def current_user
        strong_memoize(:current_user) do
          user = super

          if user
            ::Gitlab::Database::LoadBalancing::RackMiddleware
              .stick_or_unstick(env, :user, user.id)
          end

          user
        end
      end

      def authorization_header_valid?
        auth_header = headers['Authorization']
        return unless auth_header

        scope = ::Gitlab::Geo::JwtRequestDecoder.new(auth_header).decode.try { |x| x[:scope] }
        scope == ::Gitlab::Geo::API_SCOPE
      end

      def check_project_feature_available!(feature)
        not_found! unless user_project.feature_available?(feature)
      end

      def authorize_change_param(subject, *keys)
        keys.each do |key|
          authorize!("change_#{key}".to_sym, subject) if params.has_key?(key)
        end
      end

      def check_sha_param!(params, merge_request)
        if params[:sha] && merge_request.diff_head_sha != params[:sha]
          render_api_error!("SHA does not match HEAD of source branch: #{merge_request.diff_head_sha}", 409)
        end
      end

      # Normally, only admin users should have access to see LDAP
      # groups. However, due to the "Allow group owners to manage LDAP-related
      # group settings" setting, any group owner can sync LDAP groups with
      # their project.
      #
      # In the future, we should also check that the user has access to manage
      # a specific group so that we can use the Ability class.
      def authenticated_with_ldap_admin_access!
        authenticate!

        forbidden! unless current_user.admin? ||
            ::Gitlab::CurrentSettings.current_application_settings
              .allow_group_owners_to_manage_ldap
      end

      override :find_project!
      def find_project!(id)
        project = find_project(id)

        # CI job token authentication:
        # this method grants limited privileged for admin users
        # admin users can only access project if they are direct member
        ability = job_token_authentication? ? :build_read_project : :read_project

        if can?(current_user, ability, project)
          project
        else
          not_found!('Project')
        end
      end

      override :find_group!
      def find_group!(id)
        # CI job token authentication:
        # currently we do not allow any group access for CI job token
        if job_token_authentication?
          not_found!('Group')
        else
          super
        end
      end

      override :find_project_issue
      # rubocop: disable CodeReuse/ActiveRecord
      def find_project_issue(iid, project_id = nil)
        project = project_id ? find_project!(project_id) : user_project

        ::IssuesFinder.new(current_user, project_id: project.id).find_by!(iid: iid)
      end
      # rubocop: enable CodeReuse/ActiveRecord

      private

      override :project_finder_params_ee
      def project_finder_params_ee
        if params[:with_security_reports].present?
          { with_security_reports: true }
        else
          {}
        end
      end

      override :send_git_archive
      def send_git_archive(repository, **kwargs)
        EE::AuditEvents::RepositoryDownloadStartedAuditEventService.new(
          current_user,
          repository.project,
          ip_address
        ).for_project.security_event

        super
      end

      def private_token
        params[::APIGuard::PRIVATE_TOKEN_PARAM] || env[::APIGuard::PRIVATE_TOKEN_HEADER]
      end

      def job_token_authentication?
        initial_current_user && @current_authenticated_job.present? # rubocop:disable Gitlab/ModuleWithInstanceVariables
      end

      # Returns the job associated with the token provided for
      # authentication, if any
      def current_authenticated_job
        @current_authenticated_job
      end

      def warden
        env['warden']
      end

      # Check if the request is GET/HEAD, or if CSRF token is valid.
      def verified_request?
        ::Gitlab::RequestForgeryProtection.verified?(env)
      end

      # Check the Rails session for valid authentication details
      def find_user_from_warden
        warden.try(:authenticate) if verified_request?
      end

      def geo_token
        ::Gitlab::Geo.current_node.system_hook.token
      end

      def authorize_manage_saml!(group)
        unauthorized! unless can?(current_user, :admin_group_saml, group)
      end

      def check_group_saml_configured
        forbidden!('Group SAML not enabled.') unless ::Gitlab::Auth::GroupSaml::Config.enabled?
      end
    end
  end
end
