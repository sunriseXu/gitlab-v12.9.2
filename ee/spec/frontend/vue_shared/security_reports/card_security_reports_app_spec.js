import { mount } from '@vue/test-utils';
import { GlEmptyState } from '@gitlab/ui';
import MockAdapter from 'axios-mock-adapter';
import { TEST_HOST } from 'helpers/test_constants';

import CardSecurityDashboardApp from 'ee/vue_shared/security_reports/card_security_reports_app.vue';
import createStore from 'ee/security_dashboard/store';
import { trimText } from 'helpers/text_helper';
import axios from '~/lib/utils/axios_utils';

const vulnerabilitiesEndpoint = `${TEST_HOST}/vulnerabilities`;
const vulnerabilitiesSummaryEndpoint = `${TEST_HOST}/vulnerabilities_summary`;

describe('Card security reports app', () => {
  let wrapper;
  let mock;

  const runDate = new Date();
  runDate.setDate(runDate.getDate() - 7);

  const createComponent = props => {
    wrapper = mount(CardSecurityDashboardApp, {
      store: createStore(),
      stubs: ['security-dashboard-table'],
      propsData: {
        hasPipelineData: true,
        emptyStateIllustrationPath: `${TEST_HOST}/img`,
        securityDashboardHelpPath: `${TEST_HOST}/help_dashboard`,
        commit: {
          id: '1234adf',
          path: `${TEST_HOST}/commit`,
        },
        branch: {
          id: 'master',
          path: `${TEST_HOST}/branch`,
        },
        pipeline: {
          id: '55',
          created: runDate.toISOString(),
          path: `${TEST_HOST}/pipeline`,
        },
        triggeredBy: {
          path: `${TEST_HOST}/user`,
          avatarPath: `${TEST_HOST}/img`,
          name: 'TestUser',
        },
        project: {
          id: 123,
          name: 'my-project',
        },
        vulnerabilityFeedbackHelpPath: `${TEST_HOST}/vulnerability_feedback_help`,
        vulnerabilitiesEndpoint,
        vulnerabilitiesSummaryEndpoint,
        ...props,
      },
    });
  };

  beforeEach(() => {
    mock = new MockAdapter(axios);
    createComponent();
  });

  afterEach(() => {
    wrapper.destroy();
    mock.restore();
  });

  describe('Headline renders', () => {
    it('renders pipeline metadata information', () => {
      const element = wrapper.find('.card-header .js-security-dashboard-left');

      expect(trimText(element.text())).toBe('Pipeline #55 triggered 1 week ago by TestUser');

      const pipelineLink = element.find(`a[href="${TEST_HOST}/pipeline"]`);

      expect(pipelineLink).not.toBeNull();
      expect(pipelineLink.text()).toBe('#55');

      const userAvatarLink = element.find('a.user-avatar-link');

      expect(userAvatarLink).not.toBeNull();
      expect(userAvatarLink.attributes('href')).toBe(`${TEST_HOST}/user`);
      expect(userAvatarLink.find('img').attributes('src')).toBe(`${TEST_HOST}/img?width=24`);

      expect(userAvatarLink.text().trim()).toBe('TestUser');
    });

    it('renders branch and commit information', () => {
      const revInformation = wrapper.find('.card-header .js-security-dashboard-right');
      expect(revInformation.html()).toMatchSnapshot();
    });
  });

  describe('Dashboard renders properly', () => {
    const findDashboard = () => wrapper.find(CardSecurityDashboardApp);

    it('renders security dashboard', () => {
      const dashboard = findDashboard();
      expect(dashboard.exists()).toBe(true);
    });

    it('renders one filter less because projects filter is locked', () => {
      const dashboard = findDashboard();
      const filters = dashboard.findAll('.dashboard-filter');
      expect(filters.length).toBe(wrapper.vm.$store.state.filters.filters.length - 1);
    });
  });

  describe('Empty State renders correctly', () => {
    beforeEach(() => {
      createComponent({ hasPipelineData: false });
    });

    it('renders empty state component with correct props', () => {
      const emptyState = wrapper.find(GlEmptyState);

      expect(emptyState.exists()).toBe(true);
      expect(emptyState.props()).toMatchSnapshot();
    });
  });
});
