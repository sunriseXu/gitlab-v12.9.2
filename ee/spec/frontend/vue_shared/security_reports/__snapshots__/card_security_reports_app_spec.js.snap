// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card security reports app Empty State renders correctly renders empty state component with correct props 1`] = `
Object {
  "compact": false,
  "description": "The security dashboard displays the latest security report. Use it to find and fix vulnerabilities.",
  "primaryButtonLink": "http://test.host/help_dashboard",
  "primaryButtonText": "Learn more",
  "secondaryButtonLink": null,
  "secondaryButtonText": null,
  "svgPath": "http://test.host/img",
  "title": "Monitor vulnerabilities in your code",
}
`;

exports[`Card security reports app Headline renders renders branch and commit information 1`] = `
"<span class=\\"js-security-dashboard-right pull-right\\"><svg aria-hidden=\\"true\\" class=\\"s16 ic-branch\\"><use xlink:href=\\"#branch\\"></use></svg> <a href=\\"http://test.host/branch\\" class=\\"monospace\\">master</a> <span class=\\"text-muted prepend-left-5 append-right-5\\">Â·</span> <svg aria-hidden=\\"true\\" class=\\"s16 ic-commit\\">
  <use xlink:href=\\"#commit\\"></use>
</svg> <a href=\\"http://test.host/commit\\" class=\\"monospace\\">1234adf</a></span>"
`;
